### ISSUE TITLE
a simpler interface to git

### ISSUE THREAD

**Issue by goldingn** [ropensci/ozunconf17#16]
Version control is invaluable and git is an incredibly powerful version control tool. But it is sooo hard to learn. The basic idea and structure aren't that complicated, but its interface is [awful](https://stevebennett.me/2012/02/24/10-things-i-hate-about-git).  The terminology and commands are cryptic (and very dissimilar to R) and the messages and documentation are even worse:

```
You are in 'detached HEAD' state. You can look around...
```
```
DESCRIPTION
Updates remote refs using local refs, while sending objects necessary to complete the given refs.
```

Ugh.

One option for R users is to use the RStudio point-and-click git integration. Whilst that means users don't need to be familiar with the command line, it is missing a lot of functionality. E.g. I don't think it's possible to sync a remote up with another remote.

Users might find it easier to learn version control using a simpler and more R-ish interface to git (`grit`?). This could use [git2r](https://github.com/ropensci/git2r) under the hood, but define its own syntax for the most common commands and provide interpretable messages and helpfiles.

This package could do nice things like check whether there would be a merge conflict *before* messing with people's files, and then walk them through how to resolve a merge conflict. Everything would be done using git, so users would be able to switch back and forth to the RStudio GUI, commandline or whatever whenever they like; the package would just provide a nicer, and more R-like experience.

---

The syntax might look something like this, with commands run at the R prompt, within a git repo:

```r
changes()  # like 'git status' but with less junk
```
>```
>1 file changed since the last commit:
>
>    R/file.R:   1 line deleted 
>```
```r
changes("R/file.R")  # like 'git diff' but with less junk
```
>```
>   fun <- function (x) {
>-   stopifnot(x > 0)
>     x ^ 2
>   }
>```
```r
stage("R/file.R")  # like 'git add' which is a vague and confusing name
changes()
```
>```
>1 file changed since the last commit:
>
>  * R/file.R:   1 line deleted
>
>*file staged to commit
>```
```r
commit("remove check")
```
```r
branches()  # like 'git branch -v'
```
>```  
>  master  <- you are here
>  dev
>```
```r
remote_repos()  # like 'git remote -v'
```
>```
>origin   https://github.com/goldingn/test.git
>```
```r
push("origin")
pull()
sync()  # fetch from remotes, merge and push to remotes
```

**Comment by goldingn** [comment]
Following on from discussions in #10, a nice interface to simultaneously version control code, data and packages might look something like this:

```r
changes()
```
>```
>  1 dataset changed:
>    
>    data/observations.rds:   1.2Gb -> 1.3Gb
>  
>  
>  1 file changed:
>    
>    R/data_prep.Rmd:   3 lines added 
>  
>  
>  1 package changed:
>    
>    dplyr:   version 0.7.3 -> 0.7.4
>```

A simple function like `snapshot()` could be the basic interface to version control:
```r
snapshot("tweak for updated data")  
```
This would be like `git commit -am "message"` for files, also using some form of version control for data files (either a git commit, a pointer to a remote versioned filestore, or simply metadata including file size or a hash), and using drat or similar for packages. Users could also do `stage()` and `commit()` separately if they prefer, but this will likely be easier for most users to start with.

Simple version history for the whole project (like `git log`)
```r
history()
```
>```
>             10:03 - tweak for updated data
>  2017-10-13 09:12 - draft analysis script
>             16:23 - fix data prep typo
>             14:59 - update data
>             13:12 - first pass at data prep scripts
>             10:05 - add data files
>  2017-10-12 09:45 - set up repository
>```

Version history for a component of the project - might be useful for @rgayler's use case in #10?
```r
history("R/data_prep*")
```
>```
>  2017-10-13 10:03 - tweak for updated data
>             16:23 - fix data prep typo
>  2017-10-12 13:12 - first pass at data prep scripts
>```

Users could then `checkout()` a specific point in history and recover and re-run all of the files.

**Comment by goldingn** [comment]
(the name `control` is available on CRAN & bioconductor)

**Comment by njtierney** [comment]
Looks great!

Two quick things I want to mention briefly - @jennybc 's [githug](https://github.com/jennybc/githug) package helps run git from within R from the console - but what you are suggesting is of course quite different.

Also, the [available](https://github.com/ropenscilabs/available) (@jimhester + co) package, tells you whether a package name is available, and even if it is appropriate!

**Comment by goldingn** [comment]
Oh, githug is nice! I hadn't seen that. And yep, I've been using `available` a fair bit recently!

**Comment by timchurches** [comment]
I’m very interested in the versioned data store facility idea. Not sure if it should be part of the package proposed here, or a separate package/project. Something like the datastorr package by Rich Fitzjohn (not currently being progressed, I think), but not using GitHub releases as the data storage mechanism - that’s a bit too specific for my liking. I think that using git LFS makes the most sense, given it’s tight integration with git (pointers to large files are stored in git alongside all your other files), together with mechanisms to easily set up back-end data stores in Amazon AWS, and the Azure and Google cloud equivalents, and maybe even DropBox and AARNet Filestor+ (which I think has an API). A stretch goal would be support for symmetric or asymmetric encryption of the data files (which would scotch the ability to do diffs on data files, but needs must).

Are there any package for R that already address this, or which make use of git LFS in any way (rhetorical question)?

**Comment by goldingn** [comment]
Yeah, data versioning is definitely an obvious gap that needs filling. I think a separate package that does the core data versioning would be a good idea, it could then be used as a backend for this common interface

**Comment by goldingn** [comment]
@richfitz do you have plans to rejig the datastorr interface? It would be good if we could avoid doing redundant work at the ozunconf!

**Comment by HughParsonage** [comment]
I've used GitHub LFS with symmetric encryption via the `sodium` package. The pointers were quite difficult to use in practice, though maybe that's my own incompetence. But it did mean we could use sensitive data via GitHub. 

However, what worked better was just creating a bespoke data-only package for the data in question and mailing the tarballs whenever there was an update. The main benefits were simplicity, speed (even though we were using `fst` for the LFS workflow), and documentation of the data sets. There was actually no significant loss of latency for our purposes (the data was only updated weekly).

**Comment by dfalster** [comment]
Hi folks,
Great that you mentioned datastorr. I've been working with @richfitz on a draft paper about datastorr and data versioning. So it's not dead! If you're interested in a draft ping me (e.g. over slack) and I'll send it to you.  If people like the datastorr concept we could certainly consider putting time into it.  

BTW - data vesrioning has drifted from topic stated above -- new issue?

**Comment by dfalster** [comment]
On the question about git -- a simpler interface would certainly be helpful, though likely hard to achieve I suspect. The latest Rstudio release (1.1) has option of a terminal window built in which will help with more advanced git commands.

As a startign point, we could try to map the minimal set of operations we'd want to support?

**Comment by goldingn** [comment]
@dfalster yeah, I agree. The main challenge will be designing an interface that is easy to learn and use. I suspect the `snapshot()` function in `packrat` might be the most obvious. Though having used git for a while, its pretty hard for me to guess what does and doesn't make sense to new users!

**Comment by goldingn** [comment]
@njtierney could this have a middling tag too? Most of the work would be thinking about how beginners *expect* version control to work, and writing intelligible messages and documentation. Anyone could contribute something!
